hal-core/src/mem/page/alloc.rs

```diff
@@ -0,0 +1,571 @@
+    /// are 2x as large.
+    pub min_size: usize,
+
+    base_vaddr: AtomicUsize,
+    vm_offset: AtomicUsize,
+
+    /// Cache this so we don't have to re-evaluate it.
+    min_size_log2: usize,
+
+    /// Total size of the heap.
+    heap_size: AtomicUsize,
+
+    /// Array of free lists by "order". The order of an block is the number
+    /// of times the minimum page size must be doubled to reach that block's
+    /// size.
+    free_lists: L,
+}
+
+pub type Result<T> = core::result::Result<T, AllocErr>;
+#[derive(Debug)]
+pub struct Free {
+    magic: usize,
+    links: list::Links<Self>,
+    meta: Region,
+}
+
+// ==== impl BuddyAlloc ===
+
+impl BuddyAlloc {
+    #[cfg(not(loom))]
+    pub const fn new_default(min_size: usize) -> Self {
+        Self::new(
+            min_size,
+            // haha this is cool and fun
+            [
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+                spin::Mutex::new(List::new()),
+            ],
+        )
+    }
+}
+
+impl<L> BuddyAlloc<L> {
+    #[cfg(not(loom))]
+    pub const fn new(min_size: usize, free_lists: L) -> Self {
+        Self {
+            min_size,
+            base_vaddr: AtomicUsize::new(core::usize::MAX),
+            vm_offset: AtomicUsize::new(0),
+            min_size_log2: mycelium_util::math::usize_const_log2(min_size),
+            heap_size: AtomicUsize::new(0),
+            free_lists,
+        }
+    }
+
+    pub fn set_vm_offset(&self, offset: VAddr) {
+        self.vm_offset
+            .compare_exchange(0, offset.as_usize(), AcqRel, Acquire)
```

Comment:
This can be a `.swap` too, but given it's happening once during startup, I don't think it really matters...